pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -0,0 +1,1112 @@
+    NotFound,
+    NotFile,
+    WrongSize { actual: u64, expected: u64 },
+}
+
+impl std::fmt::Display for NeedsDownload {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            NeedsDownload::NotFound => write!(f, "file was not found"),
+            NeedsDownload::NotFile => write!(f, "path is not a file"),
+            NeedsDownload::WrongSize { actual, expected } => {
+                write!(f, "file size mismatch {actual} vs. {expected}")
+            }
+        }
+    }
+}
+
+/// Holds both Arc requriring that both components stay resident while holding this alive and no evictions
+/// nor garbage collection happens.
+#[derive(Clone)]
+pub(crate) struct ResidentLayer {
+    owner: Layer,
+    downloaded: Arc<DownloadedLayer>,
+}
+
+impl std::fmt::Display for ResidentLayer {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.owner)
+    }
+}
+
+impl std::fmt::Debug for ResidentLayer {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.owner)
+    }
+}
+
+impl ResidentLayer {
+    /// Release the eviction guard, making this back into a plain [`Layer`].
+    pub(crate) fn drop_eviction_guard(self) -> Layer {
+        self.into()
+    }
+
+    /// Loads all keys stored in the layer. Returns key, lsn and value size.
+    pub(crate) async fn load_keys(
+        &self,
+        ctx: &RequestContext,
+    ) -> anyhow::Result<Vec<DeltaEntry<'_>>> {
+        use LayerKind::*;
+
+        let owner = &self.owner.0;
+
+        match self.downloaded.get(owner).await? {
+            Delta(d) => {
+                owner
+                    .access_stats
+                    .record_access(LayerAccessKind::KeyIter, ctx);
+
+                // this is valid because the DownloadedLayer::kind is a OnceCell, not a
+                // Mutex<OnceCell>, so we cannot go and deinitialize the value with OnceCell::take
+                // while it's being held.
+                d.load_keys().await.context("Layer index is corrupted")
+            }
+            Image(_) => anyhow::bail!("cannot load_keys on a image layer"),
+        }
+    }
+
+    pub(crate) fn local_path(&self) -> &Path {
+        &self.owner.0.path
+    }
+
+    pub(crate) fn access_stats(&self) -> &LayerAccessStats {
+        self.owner.access_stats()
+    }
+}
+
+impl AsLayerDesc for ResidentLayer {
+    fn layer_desc(&self) -> &PersistentLayerDesc {
+        self.owner.layer_desc()
+    }
+}
+
+impl AsRef<Layer> for ResidentLayer {
+    fn as_ref(&self) -> &Layer {
+        &self.owner
+    }
+}
+
+/// Allow slimming down if we don't want the `2*usize` with eviction candidates?
+impl From<ResidentLayer> for Layer {
+    fn from(value: ResidentLayer) -> Self {
+        value.owner
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("Layer has been removed from LayerMap already")]
+pub(crate) struct RemovedFromLayerMap;
```

Comment:
Apparently if `thiserror::Error` is used there will be no unused warnings..? removing.

Reply:
Removed.